Replace ItemStack with ItemType

Wealthyturtle opened the pull request:
### Description
This PR replaces ItemStacks with ItemTypes in conditions, effects, and expressions where possible.

The search criteria I used was:
1. Expression contains "itemstack" or "itemstacks" only (I ignored expressions with "itemtype/itemstack")
2. Class doesn't already use ItemType in parsing, conditionals, etc. (Since if it did, then I'm probably screwing with it by modifying it)

Please help me review my changes :)

---
**Target Minecraft Versions:**     Any
**Requirements:**     None
**Related Issues:**     #2151, #2162, #2175, potentially more issues

bensku commented:
This needs extensive testing. A syntax test for each changed class would be best, but that is a lot of work. I'll try to help with this one, because IMO it is very important to get rid of ItemStacks.

FranKusmiruk commented on src/main/java/ch/njol/skript/effects/EffReplace.java:
```diff
@@ -100,10 +101,10 @@ protected void execute(final Event e) {
-					for (Integer slot : inv.all(item).keySet()){
-						inv.setItem(slot.intValue(), (ItemStack)replacement);
+			for (Inventory inv : (Inventory[]) haystack)
+				for (ItemType item : (ItemType[]) needles)
```
Adding an utility method like ItemType#replace might be something worth considering.

FranKusmiruk commented on src/main/java/ch/njol/skript/effects/EffSendBlockChange.java:
```diff
@@ -65,19 +66,21 @@
 			return;
 		for (Player player : players.getArray(e)) {
 			for (Block block : blocks.getArray(e)) {
+				ItemStack stack = as.getRandom();
```
This should be using getMaterial here.

FranKusmiruk commented on src/main/java/ch/njol/skript/expressions/ExprBookAuthor.java:
```diff
@@ -41,10 +41,10 @@
 	
 	static {
-		register(ExprBookAuthor.class, String.class, "[book] (author|writer|publisher)", "itemstack");
+		register(ExprBookAuthor.class, String.class, "[book] (author|writer|publisher)", "itemtype");
```
May as well support multiple itemtypes?

FranKusmiruk commented on src/main/java/ch/njol/skript/expressions/ExprBookAuthor.java:
```diff
@@ -56,7 +56,9 @@ protected String getPropertyName() {
 	@Override
-	public String convert(ItemStack itemStack) {
+	public String convert(ItemType item) {
+		ItemStack itemStack = item.getRandom();
```
You can do this check and get the item meta with the ItemType so there's no need to call getRandom.

FranKusmiruk commented on src/main/java/ch/njol/skript/expressions/ExprBookPages.java:
```diff
@@ -51,20 +51,25 @@
-		Skript.registerExpression(ExprBookPages.class, String.class, ExpressionType.PROPERTY, "[all] [the] [book] (pages|content) of %itemstack%", "%itemstack%'s [book] (pages|content)", "[book] page %number% of %itemstack%", "%itemstack%'s [book] page %number%");
+		Skript.registerExpression(ExprBookPages.class, String.class, ExpressionType.PROPERTY,
+				"[all] [the] [book] (pages|content) of %itemtype%",
+				"%itemtype%'s [book] (pages|content)",
```
Same as the other book expression.

FranKusmiruk commented on src/main/java/ch/njol/skript/expressions/ExprBookTitle.java:
```diff
@@ -40,10 +40,10 @@
 	
 	static {
-		register(ExprBookTitle.class, String.class, "(book name|title)", "itemstack");
+		register(ExprBookTitle.class, String.class, "(book name|title)", "itemtype");
```
Same here.

Wealthyturtle commented:
Just fixed a merge conflict... We should probably merge this quicker so that it's easier to maintain this branch :P

@bensku Since the Skript tests don't currently support events, the only classes we need to do syntax tests on are basically:
CondIsBlock, CondIsEdible, CondIsFlammable, CondIsSolid, CondIsTransparent, EffEnchant, EffHealth, EffReplace, ExprBookAuthor (?), ExprBookPages (?), ExprBookTitle (?), ExprItemAmount and ExprMaxStack, correct?

bensku commented:
@Wealthyturtle Alright, once we have those covered we can merge.

Wealthyturtle commented:
Ok, here is the comprehensive list of changes made:

# CondIsBlock
### Assertions
- Air
- Block
- Item

# CondIsEdible
### Assertions
- Air
- Block
- Non-edible item
- Edible items

# CondIsFlammable
### Assertions
- Air
- Flammable block
- Non-flammable block
- Item

# CondIsSolid
### Assertions
- Air
- Block
- Item

# CondIsTransparent
### Assertions
- Air
- Non-transparent block
### Special Note
- The Spigot transparent material list is incomplete, so some blocks that are supposed to be transparent, e.g. barriers and glass blocks are wrongly reported as being non-transparent. As such, I've removed the invalid example used in the class and added a notice on this issue.

# EffEnchant
### Assertions
- Non-existent enchant is not set
- Application of enchant
- Application of new enchant on item
- Disenchanting

# EffHealth
### Assertions (Items)
- New item durability
- Post-damage durability
- Post-repair durability
### Special Notes
- While rewriting the code for EffHealth, I noticed that the entity damage functionality of EffHealth is actually wrong (HealthUtils is calling e.damage() instead of setHealth(), as is being used for the HealthUtils heal() function - resulting in entity health not being properly updated when damaging mobs for 0.5 health).
- I also took the time to optimize HealthUtils (remove un-necessary Math2.fit() calls), as well as EffHealth (cast Object -> LivingEntity once instead of literally everywhere it's needed).
### Assertions (Mobs)
- New mob health
- Integer damage amount
- Decimal damage amount
- Overkill damage amount is rounded down
- Integer heal amount
- Decimal heal amount
- Fully heal amount is rounded down

# EffReplace
### Assertions
- Replace existing item completely
- Item amount of replaced items are preserved
- Replacing non-existent item results in no variable change
### Special Note
- For some reason looping all items in the inventory variable threw a SkriptAPIException. I removed the check for it and it seems to not break anything, although your verification would be good - @bensku (the error was just "", so I wasn't too sure on what to make of that).

# ExprBookAuthor
### Assertions
- New book's book author is not set
- Book author name setting (with Unicode)

# ExprBookPages
### Special Note
- I did not do a regression test for this because it is currently impossible to set a book's pages in a script alone currently.

# ExprBookTitle
### Assertions
- New book's title is not set
- Book title setting (with Unicode)
### Special Note
- I fixed the expression pattern failing for ``book title``, since the pattern's bracket was misplaced.

# ExprItemAmount
### Assertions
- New item's item amount is 1
- Regular item amount
- Over-capacity item amount
- Zero item amount

# ExprMaxStack
### Assertions
- Diamond sword stack size
- Bucket stack size
- Dirt stack size

bensku commented:
Looks good to me. I reorganized tests a bit, plus removed check for air blockiness (it is NOT a block on some MC versions).

Wealthyturtle commented:
Closing since this issue is merged :D
